src/ray/raylet/scheduling/cluster_resource_manager.h

```diff
@@ -96,6 +96,26 @@ class ClusterResourceManager {
                              const ResourceRequest &resource_request,
                              bool ignore_object_store_memory_requirement) const;
 
+  /// Add available resource to a given node.
+  /// Return false if such node doesn't exist.
+  bool AddNodeAvailableResources(scheduling::NodeID node_id,
+                                 const ResourceRequest &resource_request);
+
+  /// Update node available resources.
+  /// NOTE: This method only updates the existing resources of the node, and the
+  /// nonexistent resources will be filtered out, whitch is different from `UpdateNode`.
+  /// Return false if such node doesn't exist.
+  bool UpdateNodeAvailableResources(scheduling::NodeID node_id,
+                                    const rpc::ResourcesData &resource_data);
```

Comment:
Do we need these two APIs? Can the PG-related resource updating be done without this API? In long term, we are going to make pg resource updating the same as normal resource updating.

Reply:
> Do we need these two APIs? Can the PG-related resource updating be done without this API? In long term, we are going to make pg resource updating the same as normal resource updating.


Just like I wrote in the descriptions of this PR.

> In the process of refactoring, I found some problems in resource synchronization. For example, GCS will first receive the DeleteResources event (which is triggered by the ReturnResources of PlacementGroup on the raylet side through RPC), and then receive the 'UpdateFromResourceReport' triggered by syncer. I ideally wanted to keep the logic of UpdateFromResourceReport consistent with the raylet side (directly call cluster_resource_manager_.UpdateNode), but this will cause the PG resources just deleted to be added back.

I just move the write operations of resources from `gcs_resource_manager` to `cluster_resource_manager`

Reply:
might rename to `UpdateNodeAvailableResourcesIfExist`? also add a TODO says this will be removed once we have versioning.

Reply:
> I ideally wanted to keep the logic of UpdateFromResourceReport consistent with the raylet side (directly call cluster_resource_manager_.UpdateNode), but this will cause the PG resources just deleted to be added back.

Thanks for the clarification, but could you share more detail about this? Updates and deletes are now sent by syncer from GCS and adding will happen first, then deletion right? Why will it be added back?

Reply:
![image](https://user-images.githubusercontent.com/2016670/158513312-1f190b82-e07f-47bf-b8f3-4d3b67172ce6.png)
![image](https://user-images.githubusercontent.com/2016670/158513425-694d8e55-5288-432b-9351-9206cf1f6257.png)

Reply:
> > I ideally wanted to keep the logic of UpdateFromResourceReport consistent with the raylet side (directly call cluster_resource_manager_.UpdateNode), but this will cause the PG resources just deleted to be added back.
> 
> Thanks for the clarification, but could you share more detail about this? Updates and deletes are now sent by syncer from GCS and adding will happen first, then deletion right? Why will it be added back?

The above logs I dumped shows that the resource `GPU_group_2_c2617c823d7d38ee34e2fb7f33c4003c` is deleted at 11:36:52,985 and added back at 11:36:52,988

Reply:
Yeah, I know what you mean here. I think what I'm trying to figure out is why this is happening and whether we can make this better. @rkooo567  do you have some idea about this?

Reply:
Synced offline and I think it's ok for now.